### ISSUE TITLE
"Cannot read property '$size' of undefined" after call()

### ISSUE THREAD

**Issue by dallonf** [Netflix/falcor#477]
In my app, I have a `orgs.list.push` `call` route. It returns the following:

```
return [
      {path: ['orgs', 'list']},
      {path: ['orgs', 'list', 'push'], value: $atom({ id: id })}
    ];
```

Which I assume (docs on both `call` and the Router are very sparse at the moment) invalidates the entire `orgs.list` path and returns a value `{id: id}`. (I'm returning a value at the path of the method itself in order to return arbitrary data related to the operation. Is this the way you're supposed to that?)

This works pretty well, as long as there's _already_ some data under `orgs.list` in the cache. If not, it throws this error (I had to include line numbers manually due to some odd sourcemap configuration in my project):

```
TypeError: Cannot read property '$size' of undefined
  at onNode (falcor/lib/invalidate/invalidate-path-sets-as-json-dense.js:102)
  at walkPathSet (falcor/lib/walk/walk-path-set.js:109)
  at walkPathSet (falcor/lib/walk/walk-path-set.js:112)
  at invalidatePathSetsAsJsonDense (falcor/lib/invalidate/invalidate-path-sets-as-json-dense.js:45)
  at InvalidateResponse.subscribeToInvalidateResponse [as _subscribe] (falcor/lib/response/InvalidateResponse.js:36)
  at InvalidateResponse.Rx.Observable.observableProto.subscribe.observableProto.forEach (rx/dist/rx.js:1976)
  at observableProto.finally.observableProto.ensure (rx/dist/rx.js:4242)
  at tryCatcher (rx/dist/rx.js:567)
  at setDisposable [as action] (rx/dist/rx.js:5233)
  at ScheduledItem.invokeCore (rx/dist/rx.js:929)
```

(Update: I tried commenting out the `{path: ['orgs', 'list', 'push'], value: $atom({ id: id })}` line, since that's a somewhat suspect thing that I'm trying - but it didn't seem to help, so it's probably unrelated)

**Comment by quazzie** [comment]
I have the same problem...

**Comment by jhusain** [comment]
To invalidate the entire list you return a path-bound invalidation:
[
{path: ["orgs", "list"], invalidated: true }

]

**Comment by quazzie** [comment]
@jhusain But if you invalidate a path that is not cached you get the $size error.
On line 2357 in falcor.browser.js it does not check if the node exists in cache.

**Comment by trxcllnt** [comment]
@dallonf @quazzie this is addressed in an imminent PR. will update this thread when it's merged.

**Comment by quazzie** [comment]
any update on this ? I fixed it locally by checking if node !== null.

**Comment by trxcllnt** [comment]
@quazzie this should be fixed in master, but I don't think it's been published to npm. @michaelbpaulson or @sdesai can you confirm?

**Comment by sdesai** [comment]
Sorry about the delay. I just published 0.1.13, off of this commit: e318ed76d4e7748cd140113717f8d6d9d6d4a79d 

Which I believe should have the PR @trxcllnt mentions (I don't see any other @trxcllnt commits after this one, in master) 

If I can get the commit or PR, I can confirm whether it's in 0.1.13

**Comment by michaelbpaulson** [comment]
Any updates? @sdesai what should I confirm? Can we just add a unit test to confirm this behavior?

**Comment by sdesai** [comment]
I'll add a unit test and close.

**Comment by rnicholus** [comment]
I'm running into a very similar issue, with no obvious workaround. Simply put, I'm attempting to invalidate two items in the cache in the response to a `call`.

The route is: 'messagesById[{integers:messageIds}]["text", "swap"]'

The client-side invocation of the route is: `model.call(['messages', [0, 1], 'swap'])`. Note that the "messages" route references "messagesById". All that seems to be working, and my call route on "messagesByID" is hit as expected.

The response to the call is:
```javascript
return [
    {
        path: ['messagesById', firstId],
    },
    {
        path: ['messagesById', secondId],
    },
]
```

Client-side, my `call` fails, and I see the following in the console:
<img width="1023" alt="screen shot 2015-10-23 at 12 21 49 am" src="https://cloud.githubusercontent.com/assets/593312/10685161/186433ce-791c-11e5-9387-d677f88db324.png">

I simply want to invalidate these two items. It's not clear why this is failing, or how I should do this differently. The docs are still quite sparse regarding non-trivial use of Falcor.

**Comment by rnicholus** [comment]
At the top of the stack in my screenshot, [a `TypeError` is thrown in walkPath.js](https://github.com/Netflix/falcor/blob/v0.1.13/lib/get/walkPath.js#L17). At some point, while processing the response to my call, the `path` is `undefined`. I'm not certain _why_ this happens. Perhaps someone on the team can shed some light on this. Please see #585 for context.

Anyway, a quick fix for me involved adding the following line to the top of the exported `walkPath` function:
```javascript
If (!path) {
   return
}
```

I'm not certain what the consequences of this change are, but it seems to fix the problem for me. I'd love to hear feedback so this can be formally resolved, and am happy to help out any way I can.

**Comment by sdesai** [comment]
If you have an explicit failing integration/unit test that would help. There are a couple of integration test examples you could base it on. I created one for the original use case @dallonf mentions, but it's passing with the older version of falcor (0.1.12) also, so there's something about the use case the unit test isn't capturing.

**Comment by rnicholus** [comment]
@sdesai I can easily create a really simply example that reproduces the issue by branching off of [the fullstack-react repo](https://github.com/Widen/fullstack-react) I created for [the article I wrote last week](http://engineering.widen.com/blog/future-of-the-web-react-falcor/) about Falcor (among other stack elements). You'll be able to run it an reproduce locally by pulling down the repo, and running 3 simple commands. Will that work? If so, I'll create a branch shortly.

**Comment by sdesai** [comment]
Sure I can start with that. Thanks.

**Comment by rnicholus** [comment]
Cool! All set. The branch is at https://github.com/Widen/fullstack-react/tree/falcor-invalidate-paths-bug. 

1. Clone the repo
2. Follow the ["Start the app" steps in the README](https://github.com/Widen/fullstack-react/tree/falcor-invalidate-paths-bug#starting-the-app)
3. Navigate to http://localhost:9090
4. Attempt to add a new "name". 
5. With dev tools open, you'll see en error in the console. Also, name will not be added to the list.

**Comment by rnicholus** [comment]
@sdesai Were you able to fix this? If not, is there something else I can do to help?

**Comment by sdesai** [comment]
I haven't had a chance to look at it, but the couple of call related issues are the next thing I need to spend time on.

**Comment by sdesai** [comment]
Hey @rnicholus, I bumped up a couple of deps to the latest (which I just published yesterday), and it seems to work fine for me:

```
[~/dev/github/satyend/fullstack-react] (master) $ git diff
diff --git a/package.json b/package.json
index 02a6f04..2c12e21 100644
--- a/package.json
+++ b/package.json
@@ -8,10 +8,10 @@
   "dependencies": {
     "body-parser": "1.14.1",
     "express": "4.13.3",
-    "falcor": "0.1.13",
+    "falcor": "0.1.14",
     "falcor-express": "0.1.2",
     "falcor-http-datasource": "0.1.2",
-    "falcor-router": "0.2.9",
+    "falcor-router": "0.2.11",
     "react": "0.14.0",
     "react-dom": "0.14.0"
   },
```

<img width="1412" alt="screen shot 2015-10-29 at 12 54 50 pm" src="https://cloud.githubusercontent.com/assets/46587/10830385/910163d6-7e3c-11e5-991a-68518ff70aae.png">

**Comment by sdesai** [comment]
@dallonf @quazzie  Are your original issues resolved with falcor@0.1.14, falcor-router@0.2.11?

**Comment by rnicholus** [comment]
@sdesai Looks like the updates to falcor-router and falcor do _not_ resolve this issue. I updated and can confirm that the same issue exists. 

To reproduce in the sample repo I provided:
1. Make sure you are on the falcor-invalidate-paths-bug branch
2. Run `npm run webpack`
3. Run `node --harmony server`
4. Enter a name in the input box on the rendered page, and either hit enter or the "add name" button.

Notice that the page does not update w/ a new name, and there is an error in console.

**Comment by rnicholus** [comment]
...I've also updated falcor and falcor-router to the latest versions in that branch.

**Comment by sdesai** [comment]
@rnicholus - yeah, sorry my bad. I missed the switching branches step.

**Comment by sdesai** [comment]
Cool, thanks. Now that I have the repro repro'ing:

It does seem like a couple of open bugs:

1. Envelopes with an empty "paths" array are not handled correctly when returned as part of a call. 

We're attempting to make a followup get, based on this empty set of paths, and it's eventually barfing at the point you mention. It seems like we shouldn't even be attempting to make this follow up get. I'll look into a fix.

2. The router should not be treating a PathValue of { path: foo } as { path: foo, value:undefined } instead of { path: foo, invalidated: true } which it seems to be doing for the call, at least by design. This is working in your favor currently, but you'd need to change the PathValue returned by your 'names.add' route to be { path: ..., invalidated:true } for the long term so it doesn't break when this second issue is fixed.

For now, I'd recommend working around it by returning the updated value from your call (as you're doing on master), until we get issue 1 fixed.

**Comment by rnicholus** [comment]
Thanks for the detailed response!

Regarding your second point:

>you'd need to change the PathValue returned by your 'names.add' route to be { path: ..., invalidated:true } for the long term

In production, on a real app, I plan to pull in the `pathInvalidation` helper in "falcor-json-graph" and use that to mark paths as invalid. So, this:

```javascript
return [
   {
      path: ['names', data.names.length-1, 'name']
   },
   {
      path: ['names', 'length']
   }
]
```

...would become this:

```javascript
return [
   pathInvalidation(['names', data.names.length-1, 'name']),
   pathInvalidation(['names', 'length'])
]
```

Which will generate the invalidate path object mentioned in your comment. Correct?

My current workaround involves [the hack outlined a few messages back](https://github.com/Netflix/falcor/issues/477#issuecomment-150616716), which I'm maintaining in [a fork of Falcor](https://github.com/Widen/falcor) at the moment. That seems to be beneficial in cases where it is unpleasant to return all of the information about all of the changed paths.

**Comment by sdesai** [comment]
> Which will generate the invalidate path object mentioned in your comment. Correct?

Correct

**Comment by sdesai** [comment]
> That seems to be beneficial in cases where it is unpleasant to return all of the information about all of the changed paths.

Just to clarify:

If the value of a path has changed you definitely want to invalidate it (so you'd need to pass back at least that much) - but you're right the idea behind invalidation is that you can let the cache gradually populate, based on your natural app navigation patterns, after the call, as opposed to sending down all updated data.

To work around this issue, you'd just need to return a single value (for example, [names, length] would be sufficient in this case), and invalidate all the rest.

**Comment by rnicholus** [comment]
>To work around this issue, you'd just need to return a single value (for example, [names, length] would be sufficient in this case), and invalidate all the rest.

Great! Thanks for the clarification.

**Comment by michaelbpaulson** [comment]
This appears to be fixed.  @sdesai you seem to have the most context.  If you could close this if finished, that would be great.

**Comment by rnicholus** [comment]
Just to clarify, is this fixed? It looks like it may have been fixed elsewhere, but I'm not seeing any commits in the Falcor repo.

**Comment by sdesai** [comment]
Yes, it should be: https://github.com/Netflix/falcor/pull/600

Closing out.

**Comment by sdesai** [comment]
Just published falcor@0.1.15 to npm with the fix also:
https://github.com/Netflix/falcor/blob/0.x/CHANGELOG.md

**Comment by rnicholus** [comment]
Verified fixed in 0.1.15 on my end. Thanks!
